### ISSUE TITLE
[CGNS-237] CGNS fails to close all HDF5 objects when using external links.

### ISSUE THREAD

**Issue by brtnfld** [CGNS/CGNS#632]
> This issue has been imported from JIRA. Read the [original ticket here](https://cgnsorg.atlassian.net/browse/CGNS-237).

- _**Created at:**_ Wed, 10 Mar 2021 15:17:01 -0600

<p>Upon file close, CGNS uses HDF5 H5Fget_obj_count to get the number of open objects to close. However, when an object resides in an externally linked file, that object is not part of the object count, and these objects will not be included in the list of open objects to close returned by H5Fget_obj_ids. </p>

<p>These open objects will have a detrimental effect when CGNS is run in parallel and may cause the program to crash with too many open objects. </p>

<p>The other issue is HDF5 uses 0 size for external file cache size, so all external links will open and close each time they are accessed. Instead of keeping the file open, H5Pset_elink_file_cache_size can be increased to ensure the files are not closed.</p>

<p>H5F_OBJ_ALL can be passed as the file_id to get all the open objects, but it is not clear how this may affect parallel programs, as external links are not fully tested in parallel.</p>

<p>This issue is related to the HDF5 issue:<br/>
<a href="https://jira.hdfgroup.org/browse/HDFFV-11109" class="external-link" rel="nofollow noreferrer">https://jira.hdfgroup.org/browse/HDFFV-11109</a></p>

<p>The fix in CGNS is partially dependent on how 11109 is addressed.</p>

**Comment by brtnfld** [comment]
Re-verified live (2026-07-19). The elink-cache-size half of this issue is genuinely fixed (CG_CONFIG_HDF5_ELINK_CACHE_SIZE, commit b79637fc), but the HDF5-object-undercount half is confirmed still broken via an actual reproduction, not just static reading.

**Repro**: created `target.cgns` with 30 UserDefinedData_t nodes, and `main.cgns` with 30 cg_link_write() cross-file links into it. Opened `main.cgns` and traversed all 30 links (forcing HDF5 to resolve each one), then called `cg_close(main.cgns)`, then used raw HDF5 (`H5Fget_obj_count`/`H5Fget_obj_ids`) to inspect every HDF5 object still open process-wide.

Result: 30 HDF5 group objects (e.g. `/Base/UD_00/Payload` through `/Base/UD_29/Payload`) remain open and valid *after* `cg_close()` returns successfully — still attributed to `target.cgns`, which itself has no open file handle left through which to close them. Re-ran with 12 links instead of 30 — the leak count tracked exactly (12), ruling out a fixed/coincidental number.

Also confirmed `target.cgns`'s own HDF5 file handle never appears in `H5Fget_obj_count(..., H5F_OBJ_FILE)` even while `main.cgns` is still open, right after all 30 links were traversed — consistent with the elink-cache-size=0-by-default behavior (no persistent file handle retained), but the *objects* opened while resolving each link path are never released regardless.

**Root cause**: the `g_extids`/`n_extids` tracking array in `ADFH_Database_Close` (src/adfh/ADFH.c) only tracks ids opened within a file the caller explicitly opened via `ADFH_Database_Open` (looked up against `mta_root->g_files[]`). `target.cgns` was never itself opened that way here — it was only reached transparently through external-link resolution — so it's never registered in `g_files[]`, and the close-time cleanup loop can't reach the objects opened inside it. This array was added in 2013, predating this bug report, and appears to have never covered this class of object in the first place.
